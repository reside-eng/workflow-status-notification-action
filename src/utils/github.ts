import * as core from '@actions/core';
import { context } from '@actions/github';
import * as cache from '@actions/cache';
import { promises as fsp } from 'fs';
import { exec, ExecOptions } from '@actions/exec';
import { Inputs } from '../inputs';

const cachePrimaryKey = `last-run-status-${context.runId}-${Math.random()
  .toString(36)
  .substr(2, 12)}`;
const cacheRestoreKeys = [`last-run-status-${context.runId}-`];
const cachePaths = ['last-run-status'];

/**
 * Gets headRef according to context
 * @returns headRef
 */
export function getHeadRef(): string | undefined {
  if (context.payload.pull_request !== undefined) {
    return context.payload.pull_request.head.ref;
  }
  return context.ref.split('/').pop();
}

/**
 * Gets last run status from GH CLI
 * @returns status
 */
async function getStatusFromGithub() {
  const headRef = getHeadRef();

  const githubToken = core.getInput(Inputs.GithubToken);
  core.exportVariable('GITHUB_TOKEN', `${githubToken}`);

  const options: ExecOptions = {};
  let lastStatus = '';
  options.listeners = {
    stdout: (data: Buffer) => {
      lastStatus += data.toString();
    },
  };

  await exec(
    '/bin/bash',
    [
      '-c',
      `gh run list -w "${context.workflow}" | grep "${context.workflow}	${headRef}" | grep -v "completed	cancelled" | grep -v "in_progress" | head -n 1 | awk -F" " '{print $1"/"$2}'`,
    ],
    options,
  );

  core.info(`GH Found status: ${lastStatus}`);

  return lastStatus;
}

/**
 * Gets the last workflow run status
 * @returns last run status
 */
export async function getLastRunStatus(): Promise<string> {
  let lastStatus = '';

  const cacheKey = await cache.restoreCache(
    cachePaths,
    cachePrimaryKey,
    cacheRestoreKeys,
  );

  if (cacheKey) {
    try {
      lastStatus = await fsp.readFile(cachePaths[0], 'utf8');
      core.info(`Cache Found status: ${lastStatus}`);
    } catch (err) {
      lastStatus = await getStatusFromGithub();
    }
  } else {
    lastStatus = await getStatusFromGithub();
  }

  return lastStatus.trim();
}

/**
 *
 * @param status - Status to write to cache
 */
export async function writeStatusToCache(status: string): Promise<void> {
  await fsp.writeFile(cachePaths[0], `completed/${status}`, {
    encoding: 'utf8',
  });

  await cache.saveCache(cachePaths, cachePrimaryKey);
}
