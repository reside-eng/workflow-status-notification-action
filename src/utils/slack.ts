import { context } from '@actions/github';
import * as core from '@actions/core';
import got from 'got';
import { getHeadRef } from './github';

/**
 * Handles the actual sending request.
 * We're turning the https.request into a promise here for convenience
 *
 * @param webhookURL URL of the Slack channel Webhook
 * @param messageBody Payload to send to Slack
 */
export async function sendSlackMessage(
  webhookURL: string,
  messageBody: Record<string, string | number | boolean | unknown>,
): Promise<void> {
  core.info(`Message body: ${JSON.stringify(messageBody)}`);

  await got.post(webhookURL, {
    json: messageBody,
  });
}

/**
 * Prepare slack notification
 *
 * @param message contained in the Slack notification
 * @param status current status to notify
 * @returns the Slack message body
 */
export async function prepareSlackNotification(
  message: string,
  status: string,
): Promise<Record<string, string | number | boolean | unknown>> {
  const {
    runId,
    workflow,
    actor,
    serverUrl,
    eventName: event,
    repo: { owner },
  } = context;
  const color = status === 'success' ? 'good' : 'danger';
  const repository = context.repo.repo;
  const headRef = getHeadRef();
  const authorName = event === 'schedule' ? 'reside-camacho' : `${actor}`;

  const messageBody = {
    icon_emoji: ':bangbang:', // User icon, you can also use custom icons here
    attachments: [
      {
        // this defines the attachment block, allows for better layout usage
        color: `${color}`, // color of the attachments sidebar.
        author_name: authorName,
        author_link: `${serverUrl}/${authorName}`,
        author_icon: `${serverUrl}/${authorName}.png?size=32`,
        fields: [
          // actual fields
          {
            title: 'Repository',
            value: `${repository}`,
            short: true,
          },
          {
            title: 'Branch',
            value: `${headRef}`,
            short: true,
          },
          {
            title: 'Action URL',
            value: `<${serverUrl}/${owner}/${repository}/actions/runs/${runId}|${workflow}>`,
            short: true,
          },
          {
            title: 'Event',
            value: `${event}`,
            short: true,
          },
          {
            title: `${workflow} workflow ${status}`,
            value: `${message}`, // Custom value
            short: false, // long fields will be full width
          },
        ],
      },
    ],
  };
  return messageBody;
}
