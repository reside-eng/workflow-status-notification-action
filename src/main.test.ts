import * as core from '@actions/core';
import { context } from '@actions/github';
import * as cache from '@actions/cache';
import { promises as fsp } from 'fs';
import { mocked } from 'ts-jest/utils';
import got from 'got';
import nock from 'nock';
import run from './main';

jest.mock('@actions/core');
jest.mock('@actions/cache');

interface MockObj {
  inputs: Record<string, string | undefined>;
  repo: {
    owner: string;
    repo: string;
  };
  workflow: string;
  prNumber: number;
  runId: number;
  headRef: string;
  ref: string;
  eventName: string;
  actor: string;
  serverUrl: string;
}

let mock: MockObj;

const mockCore = core as jest.Mocked<typeof core>;
const mockCache = cache as jest.Mocked<typeof cache>;
jest.mock('got');
const mockGot = mocked(got);
const mockFn = jest.fn();
mockGot.post = mockFn.bind({});

const slackUrl = 'https://hooks.slack.com';
const slackPath = '/services/test/test';

nock(slackUrl).persist().post(slackPath).reply(200);

/**
 *
 */
function setupMock() {
  jest.clearAllMocks();

  mock = {
    // Default action inputs
    inputs: {
      'current-status': 'success',
      'slack-webhook': `${slackUrl}${slackPath}`,
      'github-token': `${process.env.GITHUB_TOKEN}`,
    },
    repo: {
      owner: 'reside-eng',
      repo: 'workflow-status-notification-action',
    },
    workflow: 'Failure workflow (for test purpose only)',
    prNumber: 4,
    runId: 23456,
    headRef: 'main',
    ref: 'main',
    eventName: 'pull_request',
    actor: 'workflowactor',
    serverUrl: 'https://github.com',
  };

  mockCore.getInput.mockImplementation(
    (name: string): string => mock.inputs[name] || '',
  );

  mockCache.restoreCache.mockImplementation(
    async (): Promise<string | undefined> => {
      try {
        await fsp.readFile('last-run-status', 'utf8');
        return 'test';
      } catch (err) {
        return undefined;
      }
    },
  );

  // Setting to this Github repo by default
  context.payload.repository = {
    name: mock.repo.repo,
    owner: {
      login: mock.actor,
    },
    clone_url: mock.repo,
  };
  context.workflow = mock.workflow;
  context.runId = mock.runId;
  context.payload.pull_request = {
    number: mock.prNumber,
    head: {
      ref: mock.headRef,
    },
  };
  context.ref = mock.ref;
  context.eventName = mock.eventName;
  context.actor = mock.actor;
  context.serverUrl = mock.serverUrl;
}

/**
 *
 */
async function writeStatusToCache() {
  fsp.writeFile('last-run-status', `completed/failure`, {
    encoding: 'utf8',
  });
}

/**
 *
 */
async function cleanCache() {
  fsp.unlink('last-run-status');
}

// Test using cache only to retrieve previous status (re-run workflow behavior)
describe('last run status retrieved from cache (re-run workflow behavior)', () => {
  beforeEach(() => setupMock());

  beforeAll(() => writeStatusToCache());

  afterAll(() => cleanCache());

  it('should send success notification if last run failed and current succeeded', async () => {
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn.mock.calls[0][0]).toStrictEqual(mock.inputs['slack-webhook']);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "good",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "pull_request",
                },
                Object {
                  "short": false,
                  "title": "Failure workflow (for test purpose only) workflow success",
                  "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });

  it('should not send notification if last run succeeded and current succeeded', async () => {
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(0);
  });

  it('should send failure notification if last run succeeded and current fails', async () => {
    mock.inputs['current-status'] = 'failure';

    await run();

    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn.mock.calls[0][0]).toStrictEqual(mock.inputs['slack-webhook']);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "danger",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "pull_request",
                },
                Object {
                  "short": false,
                  "title": "Failure workflow (for test purpose only) workflow failure",
                  "value": "Failure workflow (for test purpose only) workflow in workflow-status-notification-action failed.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });

  it('should not send notification if last run failed and current fails', async () => {
    mock.inputs['current-status'] = 'failure';

    await run();

    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(0);
  });

  it('not a pull_request: should send success notification if last run failed and current succeeded', async () => {
    context.payload.pull_request = undefined;
    context.eventName = 'workflow_dispatch';

    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn.mock.calls[0][0]).toStrictEqual(mock.inputs['slack-webhook']);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "good",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "workflow_dispatch",
                },
                Object {
                  "short": false,
                  "title": "Failure workflow (for test purpose only) workflow success",
                  "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });
});

// Test using GH CLI to retrieve previous status (new commit workflow behavior)
describe('last run status retrieved from GH CLI (new commit workflow behavior)', () => {
  beforeEach(() => setupMock());

  afterEach(() => cleanCache());

  it('should send success notification if last run failed and current succeeded', async () => {
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "good",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "pull_request",
                },
                Object {
                  "short": false,
                  "title": "Failure workflow (for test purpose only) workflow success",
                  "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });

  it('should not send notification if last run succeeded and current succeeded', async () => {
    context.workflow = 'Success workflow (for test purpose only)';

    await run();

    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(0);
  });

  it('should send failure notification if last run succeeded and current fails', async () => {
    context.workflow = 'Success workflow (for test purpose only)';
    mock.inputs['current-status'] = 'failure';

    await run();

    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "danger",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Success workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "pull_request",
                },
                Object {
                  "short": false,
                  "title": "Success workflow (for test purpose only) workflow failure",
                  "value": "Success workflow (for test purpose only) workflow in workflow-status-notification-action failed.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });

  it('should not send notification if last run failed and current fails', async () => {
    mock.inputs['current-status'] = 'failure';

    await run();

    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(0);
  });

  it('not a pull_request: should send success notification if last run failed and current succeeded', async () => {
    context.payload.pull_request = undefined;
    context.eventName = 'workflow_dispatch';

    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn.mock.calls.length).toStrictEqual(1);
    expect(mockFn.mock.calls[0][1]).toMatchInlineSnapshot(`
      Object {
        "json": Object {
          "attachments": Array [
            Object {
              "author_icon": "https://github.com/workflowactor.png?size=32",
              "author_link": "https://github.com/workflowactor",
              "author_name": "workflowactor",
              "color": "good",
              "fields": Array [
                Object {
                  "short": true,
                  "title": "Repository",
                  "value": "workflow-status-notification-action",
                },
                Object {
                  "short": true,
                  "title": "Branch",
                  "value": "main",
                },
                Object {
                  "short": true,
                  "title": "Action URL",
                  "value": "<https://github.com/reside-eng/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
                },
                Object {
                  "short": true,
                  "title": "Event",
                  "value": "workflow_dispatch",
                },
                Object {
                  "short": false,
                  "title": "Failure workflow (for test purpose only) workflow success",
                  "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
                },
              ],
            },
          ],
          "icon_emoji": ":bangbang:",
        },
      }
    `);
  });
});

// Test inputs format
describe('inputs format', () => {
  beforeEach(() => setupMock());

  beforeAll(() => writeStatusToCache());

  afterAll(() => cleanCache());

  it('should not fail with expected inputs format', async () => {
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(0);
    expect(mockFn).toBeCalledTimes(1);
  });

  it('should fail with wrong current status value', async () => {
    mock.inputs['current-status'] = 'notgood';
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(1);
    expect(mockFn).toBeCalledTimes(0);
  });

  it('should fail with wrong slack webhook format', async () => {
    mock.inputs['slack-webhook'] = 'htp:/hooks.slack.com/services/test/test';
    await run();
    expect(mockCore.setFailed).toHaveBeenCalledTimes(1);
    expect(mockCore.setFailed.mock.calls[0][0]).toMatchInlineSnapshot(
      `"Wrong Slack Webhook URL format"`,
    );
    expect(mockFn).toBeCalledTimes(0);
  });
});
