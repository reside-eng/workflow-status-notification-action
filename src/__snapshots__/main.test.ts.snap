// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last run status retrieved from GH CLI (new commit workflow behavior) not a pull_request: should send success notification if last run failed and current succeeded 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "good",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "workflow_dispatch",
          },
          {
            "short": false,
            "title": "Failure workflow (for test purpose only) workflow success",
            "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;

exports[`last run status retrieved from GH CLI (new commit workflow behavior) should send failure notification if last run succeeded and current fails 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "danger",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Success workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "pull_request",
          },
          {
            "short": false,
            "title": "Success workflow (for test purpose only) workflow failure",
            "value": "Success workflow (for test purpose only) workflow in workflow-status-notification-action failed.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;

exports[`last run status retrieved from GH CLI (new commit workflow behavior) should send success notification if last run failed and current succeeded 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "good",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "pull_request",
          },
          {
            "short": false,
            "title": "Failure workflow (for test purpose only) workflow success",
            "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;

exports[`last run status retrieved from cache (re-run workflow behavior) not a pull_request: should send success notification if last run failed and current succeeded 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "good",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "workflow_dispatch",
          },
          {
            "short": false,
            "title": "Failure workflow (for test purpose only) workflow success",
            "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;

exports[`last run status retrieved from cache (re-run workflow behavior) should send failure notification if last run succeeded and current fails 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "danger",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "pull_request",
          },
          {
            "short": false,
            "title": "Failure workflow (for test purpose only) workflow failure",
            "value": "Failure workflow (for test purpose only) workflow in workflow-status-notification-action failed.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;

exports[`last run status retrieved from cache (re-run workflow behavior) should send success notification if last run failed and current succeeded 1`] = `
{
  "json": {
    "attachments": [
      {
        "author_icon": "https://github.com/workflowactor.png?size=32",
        "author_link": "https://github.com/workflowactor",
        "author_name": "workflowactor",
        "color": "good",
        "fields": [
          {
            "short": true,
            "title": "Repository",
            "value": "workflow-status-notification-action",
          },
          {
            "short": true,
            "title": "Branch",
            "value": "main",
          },
          {
            "short": true,
            "title": "Action URL",
            "value": "<https://github.com/workflowactor/workflow-status-notification-action/actions/runs/23456|Failure workflow (for test purpose only)>",
          },
          {
            "short": true,
            "title": "Event",
            "value": "pull_request",
          },
          {
            "short": false,
            "title": "Failure workflow (for test purpose only) workflow success",
            "value": "Previously failing Failure workflow (for test purpose only) workflow in workflow-status-notification-action succeeded.",
          },
        ],
      },
    ],
    "icon_emoji": ":bangbang:",
  },
}
`;
